draft-ietf-tls-esni.md

```diff
@@ -994,6 +974,72 @@ The "ech_is_inner" is expected to have an empty payload. If the payload is
+server encrypts this message using a key derived from the HPKE context, as
+described below. It then constructs HelloRetryRequestOuter message with an
+"encrypted_client_hello" extension containing the ciphertext as its payload. It
+then continues the handshake by sending this message to the client.
+
+The payload of the "encrypted_client_hello" extension is constructed as follows.
+The client-facing server begins by computing `hrr_inner_key` and
+`hrr_inner_nonce` as described in {{client-hrr-accepted-ech}}. The extension
+payload is then computed as
+
+~~~
+    encrypted_hrr_inner = Seal(hrr_inner_key,
+                               hrr_inner_nonce,
+                               HelloRetryRequestOuterAAD,
+                               HelloRetryRequestInner)
+~~~
+
+where HelloRetryRequestOuterAAD is constructed as described in
+{{client-hrr-accepted-ech}}.
+
+The ciphertext `encrypted_hrr_inner` is used as the payload of the
+"encrypted_client_hello" extension. When sending this extension, the
+HelloRetryRequestOuter MAY also contain a "cookie" extension, but SHOULD NOT
+contain any other extension.
+
+After sending HelloRetryRequestOuter, the client-facing server does not repeat
+the steps in {{client-facing-server}} with the second ClientHelloOuter. Instead,
+it continues with the ECHConfig selection from the first ClientHelloOuter as
+follows (servers that handle HelloRetryRequest stateless would use the "cookie"
+extension to carry state required for resuming selection):
+
+If the client-facing server accepted ECH, it checks the second ClientHelloOuter
+also contains the "encrypted_client_hello" extension. If not, it MUST abort the
+handshake with a "missing_extension" alert. Otherwise, it checks that
+ClientECH.cipher_suite and ClientECH.config_id are unchanged, and that
+ClientECH.enc is empty. If not, it MUST abort the handshake with an
+"illegal_parameter" alert.
+
+Finally, it decrypts the new ClientECH.payload as a second message with the
+previous HPKE context:
+
+~~~
+    EncodedClientHelloInner = context.Open(ClientHelloOuterAAD,
+                                           ClientECH.payload)
+~~~
+
+ClientHelloOuterAAD is computed as described in {{authenticating-outer}}, but
+using the second ClientHelloOuter. If decryption fails, the client-facing
+server MUST abort the handshake with a "decrypt_error" alert. Otherwise, it
+reconstructs the second ClientHelloInner from the new EncodedClientHelloInner
+as described in {{encoding-inner}}, using the second ClientHelloOuter for
+any referenced extensions.
+
+The client-facing server then forwards the resulting ClientHelloInner to the
+backend server. It forwards all subsequent TLS messages between the client and
+backend server unmodified.
+
+If the client-facing server rejected ECH, or if the first ClientHello did not
+include an "encrypted_client_hello" extension, the client-facing server
+proceeds with the connection as usual.
```

Comment:
Actually, this paragraph is just supposed to say what the client-facing server does if it is completing the handshake and needs to send an HRR. I'm not so sure this paragraph is actually needed any more. I would favor deleting it, since there's not really any context for it anymore.

Reply:
Deleting sounds good to me.

Reply:
Done.